### PR TITLE
Minion-Gateway multi-tenant support...  Take 3

### PR DESCRIPTION
**Pull request by artnaseef**
… support.  Default tenant id to "opennms-prime" where unknown.  Kafka header "tenant-id"

- Kafka header "tenant-id"
- Default tenant ID = "opennms-prime"
- MinionGatewayProxy to inject header for minion-to-gateway comms in single-tenant/dev environments
- "Implicit" handling of tenant ID via GRPC Context
- Trapd configuration of Task Sets updated with Tenant ID

## Jira link(s)
- https://issues.opennms.org/browse/HS-xx

## Checklist
* [ ] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.

### CODE REVIEW

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/MonitorTaskSetService.java [hunk 1]
```diff
@@ -102,8 +102,11 @@ private void addMonitorTask(String location, MonitorType monitorType, IpInterfac
     }
 
     private void sendTaskSet(String location) {
+        String tenantId = "opennms-prime";  // TBD888: properly source the Tenant ID
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/MonitorTaskSetService.java:105]: need to retrieve tenantId from node

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/GrpcTaskSetPublisher.java [hunk 2]
```diff
@@ -44,16 +46,22 @@ public class GrpcTaskSetPublisher implements TaskSetPublisher {
     private static final Logger log = LoggerFactory.getLogger(GrpcTaskSetPublisher.class);
     private final TaskSetServiceGrpc.TaskSetServiceBlockingStub taskSetServiceStub;
 
+    private static final Metadata.Key HEADER_KE = Metadata.Key.of("tenant-id", Metadata.ASCII_STRING_MARSHALLER);
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/GrpcTaskSetPublisher.java:49]: nit: `HEADER_KEY`

File: minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/OpennmsGrpcServer.java [hunk 3]
```diff
@@ -307,7 +310,7 @@ public void setOutgoingMessageHandler(BiConsumer<Identity, StreamObserver<CloudT
 
     @Override
     public CompletableFuture<RpcResponseProto> dispatch(String location, RpcRequestProto request) {
-        StreamObserver<RpcRequestProto> rpcHandler = rpcConnectionTracker.lookupByLocationRoundRobin(location);
+        StreamObserver<RpcRequestProto> rpcHandler = rpcConnectionTracker.lookupByLocationRoundRobin("opennms-prime", location);
```
Comment (cgorantla) [inline minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/OpennmsGrpcServer.java:313]: Need to associate with an actual tenant id ?

File: minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/manager/rpc/LocationIndependentRpcClient.java [hunk 4]
```diff
@@ -168,9 +168,9 @@ private boolean sendRequest(RpcRequestProto requestProto) {
 
         // If a specific Minion weas requested, use it
         if (! Strings.isNullOrEmpty(requestProto.getSystemId())) {
-            rpcHandler = rpcConnectionTracker.lookupByMinionId(requestProto.getSystemId());
+            rpcHandler = rpcConnectionTracker.lookupByMinionId("opennms-prime", requestProto.getSystemId());
```
Comment (cgorantla) [inline minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/manager/rpc/LocationIndependentRpcClient.java:171]: same as above. Is there a way to associate actual tenant id here ?

File: minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/tasktresults/TaskResultsKafkaForwarder.java [hunk 5]
```diff
@@ -37,7 +58,42 @@ public SinkModule<Message, Message> getModule() {
...
+     *
+     * @param rawContent content to include as the message payload.
+     * @param tenantId Tenant ID to include in the message headers.
+     * @return ProducerRecord to send to Kafka.
+     */
+    private ProducerRecord<String, byte[]> formatProducerRecord(byte[] rawContent, String tenantId) {
+        List<Header> headers = new LinkedList<>();
+        headers.add(new RecordHeader(TENANT_ID_HEADER_NAME, tenantId.getBytes(StandardCharsets.UTF_8)));
+
+        return new ProducerRecord<String, byte[]>(
```
Comment (cgorantla) [inline minion-gateway/main/src/main/java/org/opennms/miniongateway/grpc/server/tasktresults/TaskResultsKafkaForwarder.java:91]: there is another way to set headers without using nulls 
```suggestion
       var producerRecord = new ProducerRecord<String, byte[]>(kafkaTopic, rawContent);
        producerRecord.headers().add(new RecordHeader(TENANT_ID, tenantId.getBytes(StandardCharsets.UTF_8)));
```

File: minion-gateway-grpc-proxy/main/src/main/java/org/opennms/miniongatewaygrpcproxy/grpc/CloudServiceProxyImpl.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
+package org.opennms.miniongatewaygrpcproxy.grpc;
```
Comment (cgorantla) [inline minion-gateway-grpc-proxy/main/src/main/java/org/opennms/miniongatewaygrpcproxy/grpc/CloudServiceProxyImpl.java:1]: missing copyright headers

File: charts/opennms/values.yaml [hunk 7]
```diff
@@ -91,6 +91,19 @@ OpenNMS:
...
+  MinionGatewayGrpcProxy:
+    ServiceName: opennms-minion-gateway-grpc-proxy
+    Image: opennms/horizon-stream-minion-gateway-grpc-proxy
+    ImagePullPolicy: IfNotPresent
+    Port: 8990
+    Replicas: 1
+    Resources:
+      Limits:
+        Cpu: "1"
+        Memory: 2050M
```
[the remark this answers is not in the document]
Comment (artnaseef) [reply charts/opennms/values.yaml:103]: I just copy-and-pasted this one.  Haven't dug into what it does, nor what it should be.

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/MonitorTaskSetService.java (hunk 1 above)
Comment (artnaseef) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/MonitorTaskSetService.java:105]: There's a lot more "sourcing of tenant-id" needed.  I added "TBD888" comments at key places where more work is needed.

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/GrpcTaskSetPublisher.java (hunk 2 above)
Comment (artnaseef) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/GrpcTaskSetPublisher.java:49]: Agreed - I copied that from earlier code and left it the same.

File: minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/OpennmsGrpcServer.java (hunk 3 above)
Comment (artnaseef) [reply minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/OpennmsGrpcServer.java]: Updated the code to populate the tenant-id impicitly using the GRPC Context

File: minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/manager/rpc/LocationIndependentRpcClient.java (hunk 4 above)
Comment (artnaseef) [reply minion-gateway/ipc-grpc-server/src/main/java/org/opennms/horizon/shared/ipc/grpc/server/manager/rpc/LocationIndependentRpcClient.java:171]: ditto

File: minion-gateway-grpc-proxy/main/src/main/java/org/opennms/miniongatewaygrpcproxy/grpc/CloudServiceProxyImpl.java (hunk 6 above)
Comment (artnaseef) [reply minion-gateway-grpc-proxy/main/src/main/java/org/opennms/miniongatewaygrpcproxy/grpc/CloudServiceProxyImpl.java:29]: added copyright headers

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/GrpcTaskSetPublisher.java (hunk 2 above)
Comment (artnaseef) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/GrpcTaskSetPublisher.java]: Renamed.

**Comment by artnaseef** [comment]
Superceded by https://github.com/OpenNMS/horizon-stream/pull/510

Will close shortly

**Pull request closed without merging**